doc_internal/fish-riir-plan.md

```diff
@@ -0,0 +1,77 @@
+These is a proposed port of fish-shell from C++ to Rust, and from CMake to cargo or related. This document is high level - see the Development Guide for more details.
+
+## Why Port
+
+- Gain access to more contributors and enable easier contributions. C++ is becoming a legacy language.
+- Free us from the annoyances of C++/CMake, and old toolchains.
+- Ensure fish continues to be perceived as modern and relevant.
+- Unlock concurrent mode (see below).
+
+## Why Rust
+
+- Rust is a systems programming language with broad platform support, a large community, and a relatively high probability of still being relevant in a decade.
+- Rust has a unique strength in its thread safety features, which is the missing piece to enable concurrent mode - see below.
+- Other languages considered:
+  - Java, Python and the scripting family are ruled out for startup latency and memory usage reasons.
+  - Go would be an awkward fit. fork is [quite the problem](https://stackoverflow.com/questions/28370646/how-do-i-fork-a-go-process/28371586#28371586) in Go.
+  - Other system languages (D, Nim, Zig...) are too niche: fewer contributors, higher risk of the language becoming irrelevant.
+
+## Risks
+
+- Large amount of work with possible introduction of new bugs.
```

Comment:
This is amazing, thanks for spearheading this work.
The plan sounds all-around sensible to me.

The port sounds like a great investment while being easier to pull off than a rewrite.
There are many fresh shells these days but I don't think they offer a comparable interactive experience (most of them focus on inventing their language).

Still, the port is a lot of work with late payoff, so it might be hard to find persistent contributors.
Somehow Rust projects seem to be doing fine in this regard.
Granted most of them are written from scratch.
One incremental port that died a slow death is https://github.com/remacs/remacs - but I don't think it's comparable to fish, it's around 6x larger, and the authors didn't have ownership of the code. IIRC they used a simple bindgen build; our toolchain seems much more powerful.

Reply:
"might be hard to find persistent contributions"

Just wanted to chime in and say that I love the fish shell and have been using it every day for about 6 or 7 years now. If this goes through, I will absolutely be contributing to the ongoing port and future features. I'm a full time Rust engineer and my main project at work includes quite a lot of C FFI, so I'm no stranger to that. I know I'm just one person, but figured I'd show support

Reply:
That remacs link is sobering, shanks for sharing that. I think the biggest difference is that emacs maintainers don't appear to have been involved, and the fork was scared of [touching parts of the codebase](https://nickdrozd.github.io/2019/04/16/remacs-faq.html#orgb137719).

fish gets a lot of mileage out of being written in C++ here: the types map better, the FFI tools are more powerful, and fish already uses techniques like RAII, scoped locks, shared_ptr etc which map cleanly to Rust. Still food for though.